### PR TITLE
Transformations

### PR DESCRIPTION
**Pull request by TamerShlash**
This pull request present some transformers:

* `KeyFilter`: takes a hash record and pushes a transformed version that only contains paris whose key is included in the set supplied when creating the transformer.
* `NativeTypeCast`: takes a hash record and pushes a transformed version where the values are cast to specific data types based on their keys, and on the map supplied when creating the transformer.
* `RenameKey`: takes a hash record and pushes a modified version where some keys are renamed.
* `Validator`: takes a hash record and validates if the values match given constraints, the constraints are applied based on the key. If validation failed, an error is logged and the record is not pushed. If the validation succeeds, the record is pushed.

### CODE REVIEW

File: lib/optimus_prime/transformations/native_type_cast.rb [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+      def transform(record)
+        record.keys.each do |field|
+          next unless @fields.include? field
+          case @type_map[field].downcase
+          when 'integer'
+            record[field] = Integer(record[field], 10)
+          when 'float'
+            record[field] = Float(record[field])
+          when 'boolean'
+            record[field] = (record[field].downcase == "true")
```
Comment (omarkhan) [inline lib/optimus_prime/transformations/native_type_cast.rb]: What if it's already a boolean?

Comment (TamerShlash) [inline lib/optimus_prime/transformations/native_type_cast.rb]: I was expecting values to be strings. But that is actually a very good point as the transformer might be used in other cases.

File: lib/optimus_prime/transformations/rename_key.rb [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+module OptimusPrime
+  module Destinations
+    class RenameKey < Destination
+      # This class accepts a hash object and replaces incorrect
+      # keys with correct ones based on the mapping hash given
+      # in the initializer.
+
+      def initialize(mapper:)
+        @fields = mapper.keys # Caching for faster performance
```
Comment (omarkhan) [inline lib/optimus_prime/transformations/rename_key.rb]: This actually has worse performance - array lookup is O(N), whereas hash lookup is O(1). You can just do `@mapper.include?(field)`

Comment (TamerShlash) [inline lib/optimus_prime/transformations/rename_key.rb]: :+1: I used this in two places, thanks!

File: lib/optimus_prime/transformers/validator.rb [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+      def write(record)
+        push(record) if valid?(record)
+      end
+
+      private
+
+      def valid?(record)
+        record.each do |field, value|
+          next unless @rules.include? field
+          next if send("#{@rules[field]['type']}_validator", value, @rules[field]['values'])
```
Comment (T-Dnzt) [inline lib/optimus_prime/transformers/validator.rb]: We need to support symbols for `type` and `values` since every key in the loaded YAML is a symbol.

Comment (TamerShlash) [inline lib/optimus_prime/transformers/validator.rb]: @T-Dnzt good point, will fix it.

**Pull request merged**